### ISSUE TITLE
Edit localization json using jspreadsheet

### ISSUE THREAD

**Issue by jozef-slezak** [iteria-poc/localization-json-jspreadsheet#1]
Goal: edit multiple localization JSON in one spreadsheet (see following screenshot) as webcomponent displayed in storybook.
![image](https://user-images.githubusercontent.com/16844103/114082555-2e182d80-98ae-11eb-8d9f-4f93450f28ac.png)

# Webcomponent
- [ ] implement as [webcomponent](https://javascript.plainenglish.io/can-you-build-web-components-with-svelte-3c8bc3c1cfd8)
- [x] languages is input attribute that contains JSON
- [ ] onchange returns modified copy of `languages` input
```
<iteria-messages-spreadsheet languages="{
   'en': {'budget': 'BUDGET', 'message2': 'MESSAGE2'}, 
   'sk': {'budget': 'ROZPOCET', 'message2': 'hlaska 2'}
}" onchange="console.log(event)">
</iteria-messages-spreadsheet>
```

# Spreadsheet
- [x] use spreadsheet editor https://bossanova.uk/jspreadsheet/v4/
- [ ] first column is always `messageID` (example values "budget", "message2")
- [ ] other columns (second, third ...) represents top level keys in languages json ("en", "sk" ...)
- [ ] number of rows depends on number of values ("budget", "message2")
- [ ] second column should be `en` if available (it is optional)
- [ ] other columns are sorted alphabetically
- [ ] columns and rows can be added/removed => fire onchange event containing modified languages json
![image](https://user-images.githubusercontent.com/16844103/114263083-c6fc9500-99e3-11eb-9505-73b312d72de4.png)
![image](https://user-images.githubusercontent.com/16844103/114263085-c9f78580-99e3-11eb-82a6-189b61f5ec1b.png)
- [ ] after editing each cell => fire onchange event containing modified languages json

# Storybook
- [ ] use webcomponent in [storybook](https://storybook.js.org/docs/web-components/get-started/introduction) (static site will be published from storybook in render.com)
- [x] implement at least two stories https://storybook.js.org/docs/react/get-started/browse-stories (first story will have "en", "sk", second story will have "en", "sk", "de")

**Comment by maheechy77** [comment]
Can you give me some details about this task -onchange returns modified copy of languages input. does it mean if i add/edit the spreadsheet it will return/save the modified json?(i am trying to do the tasks)

**Comment by jozef-slezak** [comment]
@maheechy77, exactly, whatever text you change in the jspreadsheet you need to fire an event.

**Comment by maheechy77** [comment]
what should be dynamic rows or columns?

**Comment by jozef-slezak** [comment]
@maheechy77, both rows and columns are dynamic. Maybe following colors explain this (I used the json and the spreadsheet from the issue definition above). See next comment.

**Comment by jozef-slezak** [comment]
![image](https://user-images.githubusercontent.com/16844103/114277911-845faa80-9a2d-11eb-98d3-65000a80fb34.png)

**Comment by maheechy77** [comment]
onchange shows only the changed column value. is this what you need?

**Comment by maheechy77** [comment]
I made a download button in which the data downloads as a csv file

**Comment by maheechy77** [comment]
I finished every tasks. now want to create a branch and push the code.What should I do?

**Comment by jozef-slezak** [comment]
Please, create a fork uding github ui and then a pull request
https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request-from-a-fork

Once I see a pull request then I will do code review.

Dňa ne 11. 4. 2021, 14:25 Muhtady Chy. Mahi ***@***.***>
napísal(a):

**Comment by jozef-slezak** [comment]
@maheechy77, you have implemented some of the review comments. I needed to repeat some of the review comments (see previous comments).

**Comment by jozef-slezak** [comment]
@maheechy77, we are still pretty far from readable and maintainable code. I see that you have introduced new functions. There are basically three problems:
1. you need top level functions with input arguments and return value (instead of local functions that use variables as side effects)
2. you need to use types (instead of anytypes) - compiler will infer types in many places 
3. you need to use map, filter, find... instead of for loop

**Comment by maheechy77** [comment]
Hello sir, Is it possible to talk. that will be easier for me to get all requirements. Thank you. Have a nice day.

**Comment by jozef-slezak** [comment]
Sure, I have setup google meet in 10 minutes.

**Comment by maheechy77** [comment]
I pushed Again. Please Check it and let me know.

**Comment by jozef-slezak** [comment]
@maheechy77, @lukasssimon, following requirement from the description of this issue are not fulfilled (after one week):

MAJOR:
* first column is always messageID (example values "budget", "message2")
* other columns (second, third ...) represents top level keys in languages json ("en", "sk" ...)
* number of rows depends on number of values ("budget", "message2")
* second column should be en if available (it is optional)
* other columns are sorted alphabetically
* columns and rows can be added/removed => fire onchange event containing modified languages json
* after editing each cell => fire onchange event containing modified languages json

MINOR:
* implement as webcomponent
* onchange returns modified copy of languages input

**Comment by maheechy77** [comment]
after editing each cell => fire onchange event containing modified languages json. Does this will also modify the props given.languages="{
   'en': {'budget': 'BUDGET', 'message2': 'MESSAGE2'}, 
   'sk': {'budget': 'ROZPOCET', 'message2': 'hlaska 2'}
}"
like if you change BUDGET to BUDGET2 it will save and become this
languages="{
   'en': {'budget': 'BUDGET2', 'message2': 'MESSAGE2'}, 
   'sk': {'budget': 'ROZPOCET', 'message2': 'hlaska 2'}
}"

**Comment by jozef-slezak** [comment]
@maheechy77, @lukasssimon, following requirement from the description of this issue are not fulfilled.

MAJOR:
- first column is always messageID (example values "budget", "message2")
- other columns (second, third ...) represents top level keys in languages json ("en", "sk" ...)
- number of rows depends on number of values ("budget", "message2")
- second column should be en if available (it is optional)
- other columns are sorted alphabetically
- columns and rows can be added/removed => fire onchange event containing modified languages json
- after editing each cell => fire onchange event containing modified languages json


MINOR:
- implement as webcomponent
- onchange returns modified copy of languages input

After too many days:
- you fulfilled only 25% of requirements (I have checked the fulfilled task in the issue description)
- the implementation of simple mapping between data structures  [`mapLangToSheetData `](https://github.com/iteria-poc/localization-json-jspreadsheet/blob/4c2111bfc7f8268235fb116085b21ce29b60f28a/src/webcomponents/helperfunctions/jsonToCol.ts#L12) is still wrong
